### PR TITLE
[CALCITE-3458] Remove desc in AbstractRelNode

### PR DESCRIPTION
**Pull request by hsyuan**
If the query is super large, e.g. contains tens of thousands of nodes or
expressions, the RelNode digest and desc become very large. The content of desc
and digest are almost the same, except that desc consists of id plus digest. So
remove desc, just use id + digest to produce description.
https://issues.apache.org/jira/browse/CALCITE-3458

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/AbstractRelNode.java [hunk 1]
```diff
@@ -363,11 +352,14 @@ public void replaceInput(
...
-    return desc;
+    return getDescription();
   }
 
+  /* Description, consists of id plus digest */
   public final String getDescription() {
-    return desc;
+    StringBuilder sb = new StringBuilder("rel#");
+    sb = sb.append(id).append(':').append(digest);
+    return sb.toString();
```
Comment (xndai) [inline core/src/main/java/org/apache/calcite/rel/AbstractRelNode.java]: If this is called before recomputeDigest(), we will get a different description string than the original implementation.  Is that you want?

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/rel/AbstractRelNode.java]: Yes. That is expected.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/AbstractRelNode.java]: I'm wondering if we can add a benchmark test for the toString with/without the fix.

File: core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatch.java [hunk 2]
```diff
@@ -144,7 +144,10 @@ private String computeDigest() {
       if (i > 0) {
         buf.append(", ");
       }
-      buf.append(rels[i].toString());
+      // Same with {@link rel#getDescription()}, we don't call it explicitly
+      // in order to avoid string copy when generating description.
+      buf.append("rel#").append(rels[i].getId())
+          .append(':').append(rels[i].getDigest());
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatch.java]: Can we abstract this description as a tool method and reuse that ? Something like `RelOptUtil#generatesDescription(StringBuilder builder, int relNodeID, String digest)`

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleMatch.java]: Sounds good. I have addressed your comments.

**Comment by chunweilei** [comment]
LGTM

**Pull request closed without merging**